test/tstTreeHelpers.cpp

```diff
@@ -0,0 +1,199 @@
+                  std::vector<int> const &parents_host,
+                  std::vector<int> const &labels_host)
+{
+  auto labels = toView<ExecutionSpace>(labels_host, "Test::labels");
+  auto parents = toView<ExecutionSpace>(parents_host, "Test::parents");
+
+  using ArborX::Experimental::reduceBVHLabels;
+  reduceBVHLabels(exec_space, parents, labels);
+
+  return Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, labels);
+}
+
+template <typename ExecutionSpace>
+auto buildRandomBVH(ExecutionSpace const &exec_space, int n)
+{
+  using MemorySpace = typename ExecutionSpace::memory_space;
+
+  Kokkos::View<ArborX::Point *, MemorySpace> points("Test::points", n);
+
+  std::default_random_engine gen;
+  std::uniform_real_distribution<float> dist(0.f, 1.f);
+  auto points_host = Kokkos::create_mirror_view(points);
+  for (int i = 0; i < n; ++i)
+    points_host(i) = ArborX::Point{dist(gen), dist(gen), dist(gen)};
+  Kokkos::deep_copy(points, points_host);
+
+  return ArborX::BVH<MemorySpace>(exec_space, points);
+}
+
+} // namespace Test
+
+#define ARBORX_TEST_REDUCE_LABELS(exec_space, parents, labels, ref)            \
+  BOOST_TEST(Test::reduceLabels(exec_space, parents, labels) == ref,           \
+             boost::test_tools::per_element())
+
+BOOST_AUTO_TEST_SUITE(TreeHelpers)
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(reduce_labels, DeviceType, ARBORX_DEVICE_TYPES)
+{
+  /*
+      [0]------------*--------------
+                    / \
+       ------*----[3] [4]*----------
+            / \         / \
+       --*[1] [2]*--   |  [5]----*--
+        / \     / \    |        / \
+       |   |   |   |   |   --*[6]  |
+       |   |   |   |   |    / \    |
+       0   1   2   3   4   5   6   7
+  */
+  auto const parents =
+      std::vector<int>{-1, 3, 3, 0, 0, 4, 5, 1, 1, 2, 2, 4, 6, 6, 5};
+  //                    0  1  2  3  4  5  6  7  8  9 10 11 12 13 14
+  //                   [0][1][2][3][4][5][6] 0  1  2  3  4  5  6  7
+
+  using ExecutionSpace = typename DeviceType::execution_space;
+  ExecutionSpace exec_space;
+
+  ARBORX_TEST_REDUCE_LABELS(
+      exec_space, parents,
+      (std::vector<int>{0, 1, 2, 3, 4, 5, 6, 0, 0, 0, 0, 0, 0, 0, 0}),
+      (std::vector<int>{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}));
+
+  ARBORX_TEST_REDUCE_LABELS(
+      exec_space, parents,
+      (std::vector<int>{0, 1, 2, 3, 4, 5, 6, 7, 7, 7, 7, 7, 7, 7, 7}),
+      (std::vector<int>{7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7}));
+
+  ARBORX_TEST_REDUCE_LABELS(
+      exec_space, parents,
+      (std::vector<int>{0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6, 7}),
+      (std::vector<int>{-1, -1, -1, -1, -1, -1, -1, 0, 1, 2, 3, 4, 5, 6, 7}));
+
+  ARBORX_TEST_REDUCE_LABELS(
+      exec_space, parents,
+      (std::vector<int>{0, 1, 2, 3, 4, 5, 6, 0, 0, 0, 3, 4, 4, 4, 7}),
+      (std::vector<int>{-1, 0, -1, -1, -1, -1, 4, 0, 0, 0, 3, 4, 4, 4, 7}));
+}
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(find_parents, DeviceType, ARBORX_DEVICE_TYPES)
```

Comment:
I wonder if there is not too much logic in that test.
How do you feel about something like
```C++
template <class MemorySpace>
struct MockBVH
{
  Kokkos::View<Kokkos::pair<int, int> *, MemorySpace> children_;
  KOKKOS_FUNCTION auto size() const { return children_.extent(0) + 1; }
};

#define HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(MEMORY_SPACE)           \
  template <>                                                                  \
  auto ArborX::Details::HappyTreeFriends::getLeftChild<MockBVH<MEMORY_SPACE>>( \
      MockBVH<MEMORY_SPACE> const &x, int i)                                   \
  {                                                                            \
    return x.children_(i).first;                                               \
  }                                                                            \
  template <>                                                                  \
  auto                                                                         \
  ArborX::Details::HappyTreeFriends::getRightChild<MockBVH<MEMORY_SPACE>>(     \
      MockBVH<MEMORY_SPACE> const &x, int i)                                   \
  {                                                                            \
    return x.children_(i).second;                                              \
  }

HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(Kokkos::HostSpace)
// would need specialization for all spaces.  can be abused by using exec spaces directly
namespace Test
{
template <class ExecutionSpace>
auto findParents(ExecutionSpace const &exec_space,
                 std::vector<Kokkos::pair<int, int>> const &children_host)
{
  auto children = toView<ExecutionSpace>(children_host, "Test::children");

  auto const n = children.extent(0) + 1;
  Kokkos::View<int *, ExecutionSpace> parents(
      Kokkos::view_alloc(Kokkos::WithoutInitializing, "Test::parents"),
      2 * n - 1);
  Kokkos::deep_copy(parents, -1);

  using ArborX::Experimental::findBVHParents;
  findBVHParents(exec_space, MockBVH<Kokkos::HostSpace>{children}, parents);
  return parents;
}
} // namespace Test

#define ARBORX_TEST_FIND_PARENTS(exec_space, children, ref)                    \
  BOOST_TEST(Test::findParents(exec_space, children) == ref,                   \
             boost::test_tools::per_element())

BOOST_AUTO_TEST_CASE_TEMPLATE(find_parents, DeviceType, ARBORX_DEVICE_TYPES)
{
  auto const parents =
      std::vector<int>{-1, 3, 3, 0, 0, 4, 5, 1, 1, 2, 2, 4, 6, 6, 5};
  //                    0  1  2  3  4  5  6  7  8  9 10 11 12 13 14
  //                   [0][1][2][3][4][5][6] 0  1  2  3  4  5  6  7

  auto const children = std::vector<Kokkos::pair<int, int>>{
      {3, 4}, {7, 8}, {9, 10}, {1, 2}, {11, 5}, {6, 14}, {12, 13}};
  //  [0]     [1]     [2]      [3]      [4]     [5]       [6]
  //  [3][4]   0  1    2   3   [1][2]    4 [5]  [6]  7     5   6

  using ExecutionSpace = typename DeviceType::execution_space;
  ExecutionSpace exec_space;

  ARBORX_TEST_FIND_PARENTS(exec_space, children, parents);
}